client/allocrunner/taskrunner/template/template_test.go

```diff
@@ -764,6 +767,110 @@ func TestTaskTemplateManager_Unblock_Multi_Template(t *testing.T) {
 	}
 }
 
+// TestTaskTemplateManager_FirstRender_Restored tests that a task that's been
+// restored renders and triggers its change mode if the template has changed
+func TestTaskTemplateManager_FirstRender_Restored(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
+	// Make a template that will render based on a key in Vault
+	vaultPath := "secret/data/password"
+	key := "password"
+	content := "barbaz"
+	embedded := fmt.Sprintf(`{{with secret "%s"}}{{.Data.data.%s}}{{end}}`, vaultPath, key)
+	file := "my.tmpl"
+	template := &structs.Template{
+		EmbeddedTmpl: embedded,
+		DestPath:     file,
+		ChangeMode:   structs.TemplateChangeModeRestart,
+	}
+
+	harness := newTestHarness(t, []*structs.Template{template}, false, true)
+	harness.start(t)
+	defer harness.stop()
+
+	// Ensure no unblock
+	select {
+	case <-harness.mockHooks.UnblockCh:
+		t.Fatalf("Task unblock should not have been called")
+	case <-time.After(time.Duration(1*testutil.TestMultiplier()) * time.Second):
+	}
+
+	// Write the secret to Vault
+	logical := harness.vault.Client.Logical()
+	_, err := logical.Write(vaultPath, map[string]interface{}{"data": map[string]interface{}{key: content}})
+	require.NoError(err)
+
+	// Wait for the unblock
+	select {
+	case <-harness.mockHooks.UnblockCh:
+	case <-time.After(time.Duration(5*testutil.TestMultiplier()) * time.Second):
+		t.Fatalf("Task unblock should have been called")
+	}
+
+	// Check the file is there
+	path := filepath.Join(harness.taskDir, file)
+	raw, err := ioutil.ReadFile(path)
+	if err != nil {
+		t.Fatalf("Failed to read rendered template from %q: %v", path, err)
+	}
+
+	if s := string(raw); s != content {
+		t.Fatalf("Unexpected template data; got %q, want %q", s, content)
+	}
+
+	// task is now running
+	harness.mockHooks.hasHandle = true
+
+	// simulate a client restart
+	harness.manager.Stop()
+	harness.mockHooks.UnblockCh = make(chan struct{}, 1)
+	harness.start(t)
+
+	// Wait for the unblock
+	select {
+	case <-harness.mockHooks.UnblockCh:
+	case <-time.After(time.Duration(5*testutil.TestMultiplier()) * time.Second):
+		t.Fatalf("Task unblock should have been called")
+	}
+
+	select {
+	case <-harness.mockHooks.RestartCh:
+		t.Fatalf("should not have restarted: %+v", harness.mockHooks)
+	case <-harness.mockHooks.SignalCh:
+		t.Fatalf("should not have restarted: %+v", harness.mockHooks)
+	case <-time.After(time.Duration(1*testutil.TestMultiplier()) * time.Second):
+	}
```

Comment:
👍 to testing that process doesn't get restarted if the template doesn't change.  However, I'm unclear how consul-template state management handles that?  I assumed consul-template library kept template rendition in memory to compare with re-renditions. On nomad agent restart, how does CT library refresh the latest state - does it read the template file and refresh its state?

Reply:
> On nomad agent restart, how does CT library refresh the latest state - does it read the template file and refresh its state?

Yeah, there are effectively 3 chunks of state that CT has:
* The `source` which gets reloaded on every client restart.
* The Consul/Vault contents, which gets refreshed on every client restart. That's actually the source of our problem here because for dynamic secrets that's a _write_ which creates a new lease and not a _read_. Every time this triggers, the [`RenderEvent.WouldRender`](https://pkg.go.dev/github.com/hashicorp/consul-template@v0.25.1/manager#RenderEvent) is true. And that includes a client restart.
* The `destination`: the CT library then compares the results with the current contents of the destination and the [`RenderEvent.DidRender`](https://pkg.go.dev/github.com/hashicorp/consul-template@v0.25.1/manager#RenderEvent) is set to false if it didn't actually change on-disk.

So what we're doing is differentiating between the `RenderEvent.WouldRender` and `RenderEvent.DidRender` cases to avoid doing extra restarts we don't need.

Also: this is making me realize we should clarify the behavior around changing the `source` out of band; if you do that while Nomad is running, CT doesn't see it. But then if you restart it will see the change.

Reply:
I'll have a CHANGELOG and upgrade guide docs PR to go with this (holding off until the giant docs replatform change lands). I'll update the docs on `source` in that PR.